shell/platform/embedder/embedder_backing_store.h

```diff
@@ -0,0 +1,94 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_EMBEDDER_EMBEDDER_BACKING_STORE_H_
+#define FLUTTER_SHELL_PLATFORM_EMBEDDER_EMBEDDER_BACKING_STORE_H_
+
+#include "flutter/shell/platform/embedder/embedder.h"
+
+namespace flutter {
+
+typedef struct {
+  /// The size of this struct. Must be sizeof(FlutterBackingStore).
+  size_t struct_size;
+  /// A baton that is not interpreted by the engine in any way. The embedder may
+  /// use this to associate resources that are tied to the lifecycle of the
+  /// `FlutterBackingStore`.
+  void* user_data;
+  /// Specifies the type of backing store.
+  FlutterBackingStoreType type;
+  /// Indicates if this backing store was updated since the last time it was
+  /// associated with a presented layer.
+  bool did_update;
+  union {
+    /// The description of the OpenGL backing store.
+    FlutterOpenGLBackingStore open_gl;
+    /// The description of the software backing store.
+    FlutterSoftwareBackingStore software;
+    /// The description of the software backing store.
+    FlutterSoftwareBackingStore2 software2;
+    // The description of the Metal backing store.
+    FlutterMetalBackingStore metal;
+    // The description of the Vulkan backing store.
+    FlutterVulkanBackingStore vulkan;
+  };
+} FlutterBackingStoreVersion1;
+
+typedef FlutterBackingStore FlutterBackingStoreVersion2;
+
+/// Wrapper around FlutterBackingStore that handles different versions of the
+/// struct.
+class EmbedderBackingStore {
+ public:
+  EmbedderBackingStore() {
+    // By default assume the oldest version of backing store. Embedder can
+    // override this by setting the struct_size field.
+    backing_store_ = {
+        .struct_size = sizeof(FlutterBackingStoreVersion1),
+    };
+  }
+
+  bool IsValid() const {
+    return backing_store_.struct_size == sizeof(FlutterBackingStoreVersion1) ||
+           backing_store_.struct_size == sizeof(FlutterBackingStoreVersion2);
```

Comment:
I think even the new backing store case is tricky since once there are fields after it, it's possible for two versions to have the same number of bytes. We'd really need an explicit version field to be safe.

I think the cumulative level of danger of accidentally breaking things is a bit scary here -- people touching FlutterMetalBackingStore likely aren't thinking about the effects their change are likely to have in other places in the embedder API and ideally people reviewing shouldn't need to keep a pile of edge-cases paged into their brains. We're looking at adding tests that would watch for such breakages in the meantime. The original API here (and a few other places with nested structs/unions) is pretty unfortunate.

Reply:
No argument here. 
Conceptually this is also problematic because the backing store is populated by client and then modified by engine, which just seems wrong.